src/bvh/iter.rs

```diff
@@ -0,0 +1,377 @@
+use crate::aabb::Bounded;
+use crate::bvh::{BVHNode, BVH};
+use crate::ray::Ray;
+
+/// Iterator traverse a BVH without memory allocations
+pub struct BVHIterator<'a, Shape: Bounded> {
+    bvh: &'a BVH,
+    ray: &'a Ray,
+    shapes: &'a [Shape],
+    stack: [usize; 32],
+    node_index: usize,
+    stack_size: usize,
+    has_node: bool,
+}
+
+impl<'a, Shape: Bounded> BVHIterator<'a, Shape> {
+    /// Creates a new BVHIterator
+    pub fn new(bvh: &'a BVH, ray: &'a Ray, shapes: &'a [Shape]) -> Self {
+        BVHIterator {
+            /// Reference to the BVH to traverse
+            bvh: bvh,
+
+            /// Reference to the input ray
+            ray: ray,
+
+            /// Reference to the input shapes array
+            shapes: shapes,
+
+            /// Traversal stack. 4 billion items seems enough?
+            stack: [0; 32],
+
+            /// Position of the iterator in bvh.nodes
+            node_index: 0,
+
+            /// Size of the traversal stack
+            stack_size: 0,
+
+            /// Whether or not we have a valid node (or leaf)
+            has_node: true,
+        }
+    }
+
+    /// Test if stack is empty.
+    fn is_stack_empty(&self) -> bool {
+        return self.stack_size == 0;
+    }
+
+    /// Push node onto stack. Not guarded against overflow.
```

Comment:
Please document panic behavior (here and other overflowing methods) as per https://rust-lang.github.io/api-guidelines/documentation.html#function-docs-include-error-panic-and-safety-considerations-c-failure

I know this is not a public function but it will still help people reading this code.

At least I think an overflow will panic in Rust. To be honest, I'm not sure, it might just wrap. Something to investigate.

Reply:
So underflowing stack_size will panic: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f8f26d84ff38a9f4c111abb9f05911a

I think an "overflow" in this case (stack_size >= 32) will just trigger an OOB panic.

I'll add the comment on both functions.